File: nextjs_app/lib/schemas/regime.ts
import { z } from "zod";

export const regimeModelVersionCreateSchema = z.object({
  name: z.string().min(1),
  course_source: z.string().default("close"),
  zscore_length: z.number().int().min(2).default(75),
  hma_length: z.number().int().min(1).default(2),
  confirmation_candles: z.number().int().min(1).default(2),
  upper_threshold: z.number(),
  lower_threshold: z.number(),
});

export const regimeModelVersionReadSchema = z.object({
  id: z.string(),
  name: z.string(),
  course_source: z.string(),
  zscore_length: z.number().int(),
  hma_length: z.number().int(),
  confirmation_candles: z.number().int(),
  upper_threshold: z.number(),
  lower_threshold: z.number(),
  created_at: z.string(),
});

export const regimeSeriesReadSchema = z.object({
  id: z.string(),
  provider_symbol: z.string(),
  asset: z.string(),
  timeframe: z.string(),
  model_version_id: z.string(),
  model_version_name: z.string().nullable(),
  period_start: z.string().nullable(),
  period_end: z.string().nullable(),
  bar_count: z.number().int(),
  unavailable_count: z.number().int(),
  last_refreshed_at: z.string().nullable(),
});

export const regimeCoverageIssueSchema = z.object({
  issue_type: z.string(),
  detail: z.string(),
});

export const regimeBarReadSchema = z.object({
  bar_time: z.string(),
  regime: z.string(),
});

export const regimeSeriesDetailReadSchema = regimeSeriesReadSchema.extend({
  bars: z.array(regimeBarReadSchema).nullable(),
  coverage_issues: z.array(regimeCoverageIssueSchema),
});

export const resultTradeReadSchema = z.object({
  id: z.string(),
  hal_result_id: z.string(),
  direction: z.string(),
  entry_time: z.string(),
  exit_time: z.string(),
  net_pnl: z.number(),
  data_source: z.string(),
  created_at: z.string(),
});

export const fetchTradesResponseSchema = z.object({
  hal_result_id: z.string(),
  trades_count: z.number().int(),
});

export const regimeDetailRowSchema = z.object({
  regime: z.string(),
  trade_count: z.number().int(),
  net_pnl: z.number(),
  max_drawdown_pct: z.number().nullable(),
  pnl_share_pct: z.number(),
  calmar_ratio: z.number().nullable(),
  sortino_ratio: z.number().nullable(),
  small_sample: z.boolean(),
});

export const regimeEvaluationReadSchema = z.object({
  id: z.string(),
  hal_result_id: z.string(),
  series_id: z.string(),
  model_version_id: z.string(),
  model_version_name: z.string().nullable(),
  coverage_pct: z.number(),
  assignment_rule: z.string(),
  is_incomplete: z.boolean(),
  total_result_pnl: z.number(),
  regime_details: z.array(regimeDetailRowSchema),
  regime_dominance: z.string().nullable(),
  created_at: z.string(),
});

export const regimeImportResponseSchema = z.object({
  series_id: z.string(),
  bars_inserted: z.number().int(),
  bars_skipped: z.number().int(),
});

export type RegimeModelVersionCreate = z.infer<typeof regimeModelVersionCreateSchema>;
export type RegimeModelVersionRead = z.infer<typeof regimeModelVersionReadSchema>;
export type RegimeSeriesRead = z.infer<typeof regimeSeriesReadSchema>;
export type RegimeSeriesDetailRead = z.infer<typeof regimeSeriesDetailReadSchema>;
export type RegimeCoverageIssue = z.infer<typeof regimeCoverageIssueSchema>;
export type RegimeBarRead = z.infer<typeof regimeBarReadSchema>;
export type ResultTradeRead = z.infer<typeof resultTradeReadSchema>;
export type FetchTradesResponse = z.infer<typeof fetchTradesResponseSchema>;
export type RegimeDetailRow = z.infer<typeof regimeDetailRowSchema>;
export type RegimeEvaluationRead = z.infer<typeof regimeEvaluationReadSchema>;
export type RegimeImportResponse = z.infer<typeof regimeImportResponseSchema>;

export const REGIME_LABELS: Record<string, string> = {
  bullish: "Bullish",
  bearish: "Bearish",
  sideways: "Seitwärts",
  "ohne Regimezuordnung": "Ohne Regimezuordnung",
};

export const REGIME_BADGE_VARIANT: Record<string, "default" | "secondary" | "destructive" | "outline"> = {
  bullish: "default",
  bearish: "destructive",
  sideways: "secondary",
  "ohne Regimezuordnung": "outline",
};

export const ISSUE_TYPE_LABELS: Record<string, string> = {
  gap: "Lücke",
  overlapping_version: "Überlappende Version",
  timeframe_mismatch: "Timeframe-Fehlanpassung",
};

File: backend/app/schemas/regime.py
"""PROJ-22: Schemas für Regime-Analyse — Modellversionen, Zeitreihen, Trades, Auswertungen."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Modellversion ──────────────────────────────────────────────────────────

class RegimeModelVersionCreate(BaseModel):
    name: str
    course_source: str = "close"
    zscore_length: int = Field(default=75, ge=2)
    hma_length: int = Field(default=2, ge=1)
    confirmation_candles: int = Field(default=2, ge=1)
    upper_threshold: float
    lower_threshold: float


class RegimeModelVersionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    course_source: str
    zscore_length: int
    hma_length: int
    confirmation_candles: int
    upper_threshold: float
    lower_threshold: float
    created_at: datetime


# ── Zeitreihe ──────────────────────────────────────────────────────────────

class RegimeSeriesRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    provider_symbol: str
    asset: str
    timeframe: str
    model_version_id: UUID
    model_version_name: str | None = None
    period_start: datetime | None = None
    period_end: datetime | None = None
    bar_count: int = 0
    unavailable_count: int = 0
    last_refreshed_at: datetime | None = None


class RegimeBarRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    bar_time: datetime
    regime: str


# ── Import ─────────────────────────────────────────────────────────────────

class RegimeImportRequest(BaseModel):
    asset: str
    timeframe: str
    model_version_id: UUID
    bars: list[RegimeBarRead]


class RegimeImportResponse(BaseModel):
    series_id: UUID
    bars_inserted: int
    bars_skipped: int


class RegimeCoverageIssue(BaseModel):
    issue_type: str   # gap, overlapping_version, timeframe_mismatch
    detail: str


class RegimeSeriesDetailRead(RegimeSeriesRead):
    bars: list[RegimeBarRead] | None = None
    coverage_issues: list[RegimeCoverageIssue] = []


# ── Trades ─────────────────────────────────────────────────────────────────

class ResultTradeRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    hal_result_id: UUID
    direction: str
    entry_time: datetime
    exit_time: datetime
    net_pnl: float
    data_source: str
    created_at: datetime


class FetchTradesResponse(BaseModel):
    hal_result_id: UUID
    trades_count: int


# ── Regime-Auswertung ──────────────────────────────────────────────────────

class RegimeDetailRow(BaseModel):
    regime: str
    trade_count: int
    net_pnl: float
    max_drawdown_pct: float | None = None
    pnl_share_pct: float
    calmar_ratio: float | None = None
    sortino_ratio: float | None = None
    small_sample: bool = False


class RegimeEvaluationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    hal_result_id: UUID
    series_id: UUID
    model_version_id: UUID
    model_version_name: str | None = None
    coverage_pct: float
    assignment_rule: str
    is_incomplete: bool
    total_result_pnl: float
    regime_details: list[RegimeDetailRow]
    regime_dominance: str | None = None  # name of dominant regime or None
    created_at: datetime
